src/Hedgehog/Internal/Function.idr

```diff
@@ -0,0 +1,351 @@
+      cogenImplDef fun impl = def impl $ pure $ patClause (var impl) (var "MkCogen" `app` var fun)
+
+
+  ||| Alias for `CogenVis Public`
+  export %inline
+  Cogen : List Name -> ParamTypeInfo -> Res (List TopLevel)
+  Cogen = CogenVis Public
+
+||| Generates a random function being given a generator of codomain type
+|||
+||| This function takes a co-generator of domain type using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
+||| If you need a showable function, you have to use a shrinkable version,
+||| which requires more strict implementation on the domain type.
+export
+function_ : Cogen a => Gen b -> Gen (a -> b)
+function_ bg = MkGen $ \sz, sd => singleton $ \x => value $ unGen bg sz $ perturb x sd
+
+||| Generates a random dependently typed function being given a generator of codomain type family
+|||
+||| This function takes a co-generator of domain type using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
+export
+depfun_ : Cogen a => {0 b : a -> Type} -> ((x : a) -> Gen $ b x) -> Gen ((x : a) -> b x)
+depfun_ bg = MkGen $ \sz, sd => singleton $ \x => value $ unGen (bg x) sz $ perturb x sd
+
+||| Generates a random function with dependently typed domain being given a generator of codomain type
+|||
+||| This function takes a co-generator of domain type family using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
+export
+dargfun_ : {0 b : a -> Type} -> ({0 x : a} -> Cogen (b x)) => Gen c -> Gen ({0 x : a} -> b x -> c)
+dargfun_ bg = MkGen $ \sz, sd => singleton $ \x => value $ unGen bg sz $ perturb x sd
+
+||| Generates a random dependently typed function with dependently typed domain
+||| being given a generator of codomain type family
+|||
+||| This function takes a co-generator of domain type family using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
+export
+dargdepfun_ : {0 b : a -> Type} ->
+              {0 c : {0 x : a} -> b x -> Type} ->
+              ({0 x : a} -> Cogen (b x)) =>
+              ({0 x : a} -> (y : b x) -> Gen (c y)) -> Gen ({0 x : a} -> (y : b x) -> c y)
+dargdepfun_ bg = MkGen $ \sz, sd => singleton $ \x => value $ unGen (bg x) sz $ perturb x sd
+
+----------------------------
+--- Shrinkable functions ---
+----------------------------
+
+-- Claessen, K. (2012, September). Shrinking and showing functions:(functional pearl).
+-- In ACM SIGPLAN Notices (Vol. 47, No. 12, pp. 73-80). ACM.
+
+infixr 5 :->
+
+public export
+data (:->) : Type -> Type -> Type where
```

Comment:
This needs thorough documentation. Please add the link from above to the doc string but also explain in some details what this does and why it is needed.

Also, please consider giving this an identifier name instead of an operator. Otherwise, please explain why you think an operator is preferable in this case.

Reply:
I've added some, please take a look.

Speaking of an operator -- well, as almost always, operators are for nice look in signatures. Since this type is for specialised form of functions, it's nice to use an arrow for them. Moreover, this arrow is the same as in Haskell's Hedgehog and Quickcheck, thus, people from there will feel familiar with this operator.

Reply:
This is really good now! Thanks for doing this thoroughly. It helps a lot with understanding what's going on and also with later maintaining the code.